src/ILCompiler.WebAssembly/src/Compiler/WebAssemblyCodegenCompilation.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+
+using Internal.TypeSystem;
+
+using ILCompiler.DependencyAnalysis;
+using ILCompiler.DependencyAnalysisFramework;
+using LLVMSharp;
+
+namespace ILCompiler
+{
+    public sealed class WebAssemblyCodegenCompilation : Compilation
+    {
+        internal WebAssemblyCodegenConfigProvider Options { get; }
+        internal LLVMModuleRef Module { get; }
+
+        internal WebAssemblyCodegenCompilation(
+            DependencyAnalyzerBase<NodeFactory> dependencyGraph,
+            NodeFactory nodeFactory,
+            IEnumerable<ICompilationRootProvider> roots,
+            Logger logger,
+            WebAssemblyCodegenConfigProvider options)
+            : base(dependencyGraph, nodeFactory, GetCompilationRoots(roots, nodeFactory), null, logger)
+        {
+            LLVM.LoadLibrary_libLLVM("./libLLVM-x64.dll");
+            Module = LLVM.ModuleCreateWithName("netscripten");
+            LLVM.SetTarget(Module, "asmjs-unknown-emscripten");
+            Options = options;
+        }
+
+        private static IEnumerable<ICompilationRootProvider> GetCompilationRoots(IEnumerable<ICompilationRootProvider> existingRoots, NodeFactory factory)
+        {
+            yield return new CppCodegenCompilationRootProvider(factory.TypeSystemContext);
+
+            foreach (var existingRoot in existingRoots)
+                yield return existingRoot;
+        }
+
+        protected override void CompileInternal(string outputFile, ObjectDumper dumper)
+        {
+            _dependencyGraph.ComputeMarkedNodes();
+
+            var nodes = _dependencyGraph.MarkedNodeList;
+
+            WebAssemblyObjectWriter.EmitObject(outputFile, nodes, NodeFactory, this, dumper);
+        }
+
+        protected override void ComputeDependencyNodeDependencies(List<DependencyNodeCore<NodeFactory>> obj)
+        {
+            foreach (WebAssemblyMethodCodeNode methodCodeNodeNeedingCode in obj)
+            {
+                Internal.IL.ILImporter.CompileMethod(this, methodCodeNodeNeedingCode);
+            }
+        }
+
+        private class CppCodegenCompilationRootProvider : ICompilationRootProvider
```

Comment:
Maybe this is not needed.

Reply:
@davidwrighton , might we need this (albeit renamed)?

Reply:
Well, we'll eventually likely need something similar, but we don't look to need it now, and I don't think it should be in this file when we finally do need it.

Reply:
Removed.